x/programs/cmd/simulator/src/lib.rs

```diff
@@ -223,109 +225,99 @@ impl Client {
+            .arg("interpreter")
+            .arg("--cleanup")
+            .arg("--log-level")
+            .arg("error")
+            .stdin(Stdio::piped())
+            .stdout(Stdio::piped())
+            .spawn()?;
+
+        let (Some(stdin), Some(stdout)) = (child.stdin.as_mut(), child.stdout.as_mut()) else {
+            panic!("could not attach to the child process stdio");
         };
+        let reader = BufReader::new(stdout);
+        let lines = &mut reader.lines();
 
-        run_step(self.path, plan)
-    }
-}
+        let _ = Self::read_response(lines)?; // intepreter ready
 
-fn cmd_output<P>(path: P, plan: &Plan) -> Result<Output, Box<dyn Error>>
-where
-    P: AsRef<OsStr>,
-{
-    let mut child = Command::new(path)
-        .arg("run")
-        .arg("--cleanup")
-        .arg("-")
-        .stdin(Stdio::piped())
-        .stdout(Stdio::piped())
-        .spawn()?;
-
-    // write json to stdin
-    let input =
-        serde_json::to_string(plan).map_err(|e| format!("failed to serialize json: {e}"))?;
-    if let Some(ref mut stdin) = child.stdin {
-        stdin
-            .write_all(input.as_bytes())
-            .map_err(|e| format!("failed to write to stdin: {e}"))?;
-    }
+        let res = Self::run_steps(stdin, lines, plan)?;
 
-    child
-        .wait_with_output()
-        .map_err(|e| format!("failed to wait for child: {e}").into())
-}
+        // kill the process to avoid it to read an EOF from stdin
+        child.kill().unwrap();
 
-fn run_steps<P, T>(path: P, plan: &Plan) -> Result<Vec<T>, Box<dyn Error>>
-where
-    T: serde::de::DeserializeOwned + serde::Serialize,
-    P: AsRef<OsStr>,
-{
-    let output = cmd_output(path, plan)?;
-    let mut items: Vec<T> = Vec::new();
-
-    if !output.status.success() {
-        println!("stderr");
-        for line in String::from_utf8_lossy(&output.stderr).lines() {
-            println!("{line}");
-        }
-        println!("stdout");
-        for line in String::from_utf8_lossy(&output.stdout).lines() {
-            println!("{line}");
-        }
-        return Err(String::from_utf8(output.stdout)?.into());
+        Ok(res)
     }
 
-    for line in String::from_utf8(output.stdout)?
-        .lines()
-        .filter(|line| !line.trim().is_empty())
-    {
-        let item = serde_json::from_str(line)
-            .map_err(|e| format!("failed to parse output to json: {e}"))?;
-        items.push(item);
+    fn run_steps(
+        stdin: &mut ChildStdin,
+        lines: &mut Lines<BufReader<&mut ChildStdout>>,
+        plan: &Plan,
+    ) -> Result<Vec<PlanResponse>, Box<dyn Error>> {
+        let mut items: Vec<PlanResponse> = Vec::new();
+        for step in &plan.steps {
+            let resp = Self::run_step(stdin, lines, plan.caller_key.clone(), step.clone())?;
+            items.push(resp);
+        }
+        Ok(items)
     }
 
-    Ok(items)
+    fn run_step(
+        stdin: &mut ChildStdin,
+        lines: &mut Lines<BufReader<&mut ChildStdout>>,
```

Comment:
When you're always passing in the things, this should be a good hint that you should just create a new type with methods like `write`, `run_step`, and `run_plan`.

Reply:
Also, you shouldn't need to use this `Lines` type directly. You should be dealing with `Iterator<Item = Result<Resposne, Error>>` instead.

Reply:
I think that we could just refactor the `Client` a bit to do that and have the over-used parameters as fields of the struct

Reply:
Yeah you're right, a client is should not be tied to a simulator process. So I will create a new struct that denotes a single simulator process.

Reply:
updated by creating a `SimulatorChild` which manages all that.